### PR TITLE
Issue/138 save if task is done by a user

### LINKED ISSUE #138: Save if task is done by a user
# Description

# Suggestion

### PR DESCRIPTION
**Pull request by msentwickler**
Closes #138 

## Related Issue
https://github.com/E-Edu/task/issues/138

## Type of change

- [ ] Bug fix
- [x] Feature request
- [ ] Documenting code

## Checklists

- [x] I referenced the according Issue
  - [ ] I created an Issue
  - [x] I put a link in the Section [Related Issue](#related-issue)
- [ ] I tested my code with multiple browsers
- [x] My commit messages are good and understandable

### CODE REVIEW

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 1]
```diff
@@ -89,12 +96,16 @@
...
 
-	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey) {
+	public ControllerResult<List<TaskModel>> getTasksFromLecture(String lectureNameKey, boolean showBanned) {
 		ControllerResult<LectureModel> lectureResult = this.lectureController
 			.getLectureByNameKey(lectureNameKey);
 		if (lectureResult.isResultNotPresent()) return ControllerResult.of(Error.NOT_FOUND, Constants.Lecture.NAME_KEY);
 
-		return ControllerResult.of(this.taskRepository.getAllTasksByLectureId(lectureResult.getResult()));
+		List<TaskModel> tasks = this.taskRepository.getAllTasksByLectureId(lectureResult.getResult());
+		tasks.removeIf(tm -> userBanRepository.existsByUserId(UUID.randomUUID()) || userBanRepository.existsByUserId(UUID.randomUUID()) && !showBanned);
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java:105]: why two times?

File: src/main/java/de/themorpheus/edu/taskservice/database/model/TaskDoneModel.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+@Entity
+@Table
+public class TaskDoneModel {
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private int taskDoneId;
+
+	private UUID userId;
+	private int taskId;
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/database/model/TaskDoneModel.java:26]: Use TaskModel

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java [hunk 3]
```diff
@@ -76,4 +76,9 @@ public Object getTask(@PathVariable @Min(1) int taskId) {
 		return this.taskController.getTaskByTaskId(taskId).getHttpResponse();
 	}
 
+	@PostMapping("/task/done")
+	public Object markAsDone(@RequestBody @Valid MarkTaskAsDoneDTO dto) {
+		return this.taskController.markTaskAsDone(dto.getTaskId(), dto.getDate()).getHttpResponse();
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java:81]: Use timestamp and generate it in the Controller

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/dto/request/MarkTaskAsDoneDTO.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package de.themorpheus.edu.taskservice.endpoint.dto.request;
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/dto/request/MarkTaskAsDoneDTO.java:1]: Delete date and get taskId from a pathVariable

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java (hunk 3 above)
Comment (PAException) [reply src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java:81]: rename to markTaskAsDone

File: src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java [hunk 5]
```diff
@@ -36,8 +36,8 @@ public Object createTask(@RequestBody @Valid CreateTaskRequestDTO dto) {
 	}
 
 	@GetMapping("/lecture/{lectureNameKey}/task")
-	public Object getAllTasksFromLecture(@PathVariable @NotNull @NotEmpty @NotBlank String lectureNameKey) {
-		return this.taskController.getTasksFromLecture(lectureNameKey).getHttpResponse();
+	public Object getAllTasksFromLecture(@PathVariable @NotBlank String lectureNameKey, @RequestParam boolean showBanned) {
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/endpoint/TaskEndpoint.java:39]: [Make @ RequestParam not required](https://www.baeldung.com/spring-request-param#making-an-optional-request-parameter)

File: src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java [hunk 6]
```diff
@@ -3,17 +3,22 @@
...
 import de.themorpheus.edu.taskservice.database.repository.TaskRepository;
+import de.themorpheus.edu.taskservice.database.repository.UserBanRepository;
 import de.themorpheus.edu.taskservice.endpoint.dto.request.CreateTaskRequestDTO;
 import de.themorpheus.edu.taskservice.endpoint.dto.request.UpdateTaskRequestDTO;
 import de.themorpheus.edu.taskservice.endpoint.dto.response.GetSolutionTypeResponseDTO;
 import de.themorpheus.edu.taskservice.util.Constants;
 import de.themorpheus.edu.taskservice.util.ControllerResult;
 import de.themorpheus.edu.taskservice.util.Error;
 import de.themorpheus.edu.taskservice.util.Validation;
+
```
Comment (PAException) [inline src/main/java/de/themorpheus/edu/taskservice/controller/TaskController.java:20]: empty line

**Review by PAException — CHANGES_REQUESTED** [review]
Please also check if taskIs banned on task/next

**Comment by PAException** [comment]
And Next time create branch from experimental!!!

**Comment by msentwickler** [comment]
Created a new one with right base branch.

**Pull request closed without merging**